From 1866bf6f82acaa0bc139ff95cd4ca376e12b1be9 Mon Sep 17 00:00:00 2001
From: Jerome Martinez <jerome@mediaarea.net>
Date: Wed, 3 Sep 2025 19:54:10 +0200
Subject: [PATCH 07/19] matroskaenc: remove unused MaxBlockAdditionID

For non video tracks: it is never used, let's save 4 bytes
For video tracks: simplify the temporary storage with only a void element and MaxBlockAdditionID will be written when finishing the track if needed (not the default value)
---
 libavformat/matroskaenc.c                     |  8 ++--
 tests/ref/fate/atrac-matroska-remux           |  4 +-
 tests/ref/fate/matroska-avoid-negative-ts     |  4 +-
 tests/ref/fate/matroska-dvbsub-remux          |  4 +-
 tests/ref/fate/matroska-flac-extradata-update |  4 +-
 tests/ref/fate/matroska-move-cues-to-front    |  4 +-
 tests/ref/fate/matroska-ms-mode               |  4 +-
 tests/ref/fate/matroska-pgs-remux             |  4 +-
 tests/ref/fate/matroska-pgs-remux-durations   |  4 +-
 tests/ref/fate/matroska-qt-mode               |  4 +-
 tests/ref/fate/rgb24-mkv                      |  2 +-
 tests/ref/lavf/mka                            |  4 +-
 tests/ref/lavf/mkv                            |  4 +-
 tests/ref/lavf/mkv_attachment                 |  4 +-
 tests/ref/seek/lavf-mkv                       | 44 +++++++++----------
 15 files changed, 51 insertions(+), 51 deletions(-)

diff --git a/libavformat/matroskaenc.c b/libavformat/matroskaenc.c
index a8408da..860ef1c 100644
--- a/libavformat/matroskaenc.c
+++ b/libavformat/matroskaenc.c
@@ -1735,16 +1735,16 @@ static void mkv_write_blockadditionmapping(AVFormatContext *s, const MatroskaMux
                                            const AVCodecParameters *par, AVIOContext *pb,
                                            mkv_track *track, const AVStream *st)
 {
-    if (IS_SEEKABLE(s->pb, mkv)) {
+    if (IS_SEEKABLE(s->pb, mkv) && par->codec_type == AVMEDIA_TYPE_VIDEO) {
         track->blockadditionmapping_offset = avio_tell(pb);
         // We can't know at this point if there will be a block with BlockAdditions, so
         // we either write the default value here, or a void element. Either of them will
         // be overwritten when finishing the track.
-        put_ebml_uint(pb, MATROSKA_ID_TRACKMAXBLKADDID, 0);
         if (par->codec_type == AVMEDIA_TYPE_VIDEO) {
             // Similarly, reserve space for an eventual
             // HDR10+ ITU T.35 metadata BlockAdditionMapping.
-            put_ebml_void(pb, 3 /* BlockAdditionMapping */
+            put_ebml_void(pb, 4 /* MaxBlockAdditionID */
+                            + 3 /* BlockAdditionMapping */
                             + 4 /* BlockAddIDValue */
                             + 4 /* BlockAddIDType */);
         }
@@ -3426,7 +3426,7 @@ after_cues:
         for (unsigned i = 0; i < s->nb_streams; i++) {
             const mkv_track *track = &mkv->tracks[i];
 
-            if (!track->max_blockaddid)
+            if (!track->max_blockaddid || !track->blockadditionmapping_offset)
                 continue;
 
             // We reserved a single byte to write this value.
diff --git a/tests/ref/fate/atrac-matroska-remux b/tests/ref/fate/atrac-matroska-remux
index 1f62795..3f5974a 100644
--- a/tests/ref/fate/atrac-matroska-remux
+++ b/tests/ref/fate/atrac-matroska-remux
@@ -1,5 +1,5 @@
-1e0eab5c5cb6ee0a77ff9c4611a3044b *tests/data/fate/atrac-matroska-remux.matroska
-450932 tests/data/fate/atrac-matroska-remux.matroska
+3dfec76d1df4b69d4b8fcbf19d1fa2bc *tests/data/fate/atrac-matroska-remux.matroska
+450928 tests/data/fate/atrac-matroska-remux.matroska
 #tb 0: 1/1000
 #media_type 0: audio
 #codec_id 0: atrac1
diff --git a/tests/ref/fate/matroska-avoid-negative-ts b/tests/ref/fate/matroska-avoid-negative-ts
index 2295969..b86ccdb 100644
--- a/tests/ref/fate/matroska-avoid-negative-ts
+++ b/tests/ref/fate/matroska-avoid-negative-ts
@@ -1,5 +1,5 @@
-fb4e7a969ef65f61c4c45d5976188aa2 *tests/data/fate/matroska-avoid-negative-ts.matroska
-973080 tests/data/fate/matroska-avoid-negative-ts.matroska
+30f3bed6e6a238bdaeb0d697b17e0a2b *tests/data/fate/matroska-avoid-negative-ts.matroska
+973076 tests/data/fate/matroska-avoid-negative-ts.matroska
 #extradata 0:       22, 0x2885037c
 #tb 0: 1/1000
 #media_type 0: video
diff --git a/tests/ref/fate/matroska-dvbsub-remux b/tests/ref/fate/matroska-dvbsub-remux
index f55b964..b18ab9e 100644
--- a/tests/ref/fate/matroska-dvbsub-remux
+++ b/tests/ref/fate/matroska-dvbsub-remux
@@ -1,5 +1,5 @@
-d6385d4623cd59061c13e26b004eaa8e *tests/data/fate/matroska-dvbsub-remux.matroska
-39016 tests/data/fate/matroska-dvbsub-remux.matroska
+a05501ff60051aca59fd67b8bac69a63 *tests/data/fate/matroska-dvbsub-remux.matroska
+39008 tests/data/fate/matroska-dvbsub-remux.matroska
 #extradata 0:        5, 0x00bb0064
 #extradata 1:        5, 0x00bb0064
 #tb 0: 1/1000
diff --git a/tests/ref/fate/matroska-flac-extradata-update b/tests/ref/fate/matroska-flac-extradata-update
index 79acbb7..077e3a7 100644
--- a/tests/ref/fate/matroska-flac-extradata-update
+++ b/tests/ref/fate/matroska-flac-extradata-update
@@ -1,5 +1,5 @@
-75ec39f0faaaff953e2d46b431acc307 *tests/data/fate/matroska-flac-extradata-update.matroska
-1828 tests/data/fate/matroska-flac-extradata-update.matroska
+068409ca72dde7a059159c4b3ee98e36 *tests/data/fate/matroska-flac-extradata-update.matroska
+1816 tests/data/fate/matroska-flac-extradata-update.matroska
 #extradata 0:       34, 0x7166090c
 #extradata 1:       34, 0x7166090c
 #extradata 2:       34, 0x7166090c
diff --git a/tests/ref/fate/matroska-move-cues-to-front b/tests/ref/fate/matroska-move-cues-to-front
index 1af5b1f..8571f3b 100644
--- a/tests/ref/fate/matroska-move-cues-to-front
+++ b/tests/ref/fate/matroska-move-cues-to-front
@@ -1,5 +1,5 @@
-deeb9e63e52e015e5f7f592f3417f385 *tests/data/fate/matroska-move-cues-to-front.matroska
-23056700 tests/data/fate/matroska-move-cues-to-front.matroska
+1c62f605c71edf474eaad7004a5a02b2 *tests/data/fate/matroska-move-cues-to-front.matroska
+23056692 tests/data/fate/matroska-move-cues-to-front.matroska
 #tb 0: 1/1000
 #media_type 0: audio
 #codec_id 0: pcm_s24be
diff --git a/tests/ref/fate/matroska-ms-mode b/tests/ref/fate/matroska-ms-mode
index 4b6b0e9..bde90b7 100644
--- a/tests/ref/fate/matroska-ms-mode
+++ b/tests/ref/fate/matroska-ms-mode
@@ -1,5 +1,5 @@
-e7f44cd6a5c0f45fea11874afb8c1c0d *tests/data/fate/matroska-ms-mode.matroska
-413103 tests/data/fate/matroska-ms-mode.matroska
+13516241b74daac97fcbc070e91c7e17 *tests/data/fate/matroska-ms-mode.matroska
+413099 tests/data/fate/matroska-ms-mode.matroska
 #extradata 0:       40, 0x54290c93
 #extradata 1:      114, 0xb6c80771
 #tb 0: 1/1000
diff --git a/tests/ref/fate/matroska-pgs-remux b/tests/ref/fate/matroska-pgs-remux
index 318cc65..f73e57f 100644
--- a/tests/ref/fate/matroska-pgs-remux
+++ b/tests/ref/fate/matroska-pgs-remux
@@ -1,5 +1,5 @@
-b03d39ae755cb07fde023f6e9e6186ce *tests/data/fate/matroska-pgs-remux.matroska
-49747 tests/data/fate/matroska-pgs-remux.matroska
+8380e83f7b2874feed3c0b34e54dce6e *tests/data/fate/matroska-pgs-remux.matroska
+49743 tests/data/fate/matroska-pgs-remux.matroska
 #tb 0: 1/1000
 #media_type 0: subtitle
 #codec_id 0: hdmv_pgs_subtitle
diff --git a/tests/ref/fate/matroska-pgs-remux-durations b/tests/ref/fate/matroska-pgs-remux-durations
index 4a07f5e..2aedd42 100644
--- a/tests/ref/fate/matroska-pgs-remux-durations
+++ b/tests/ref/fate/matroska-pgs-remux-durations
@@ -1,5 +1,5 @@
-1d8fedf8f90f755f9dd6914489ee8ab2 *tests/data/fate/matroska-pgs-remux-durations.matroska
-49759 tests/data/fate/matroska-pgs-remux-durations.matroska
+52ac72c44facd2cc948637ac8528beb7 *tests/data/fate/matroska-pgs-remux-durations.matroska
+49755 tests/data/fate/matroska-pgs-remux-durations.matroska
 #tb 0: 1/1000
 #media_type 0: subtitle
 #codec_id 0: hdmv_pgs_subtitle
diff --git a/tests/ref/fate/matroska-qt-mode b/tests/ref/fate/matroska-qt-mode
index b92963f..24577f0 100644
--- a/tests/ref/fate/matroska-qt-mode
+++ b/tests/ref/fate/matroska-qt-mode
@@ -1,5 +1,5 @@
-bab6f4122a60a0d4e192fb93c17a5464 *tests/data/fate/matroska-qt-mode.matroska
-1884240 tests/data/fate/matroska-qt-mode.matroska
+399f6dc1c1b3c623e670e0dc32b29a36 *tests/data/fate/matroska-qt-mode.matroska
+1884236 tests/data/fate/matroska-qt-mode.matroska
 #extradata 0:       90, 0x817d0185
 #tb 0: 1/1000
 #media_type 0: video
diff --git a/tests/ref/fate/rgb24-mkv b/tests/ref/fate/rgb24-mkv
index 59c532b..4da78cf 100644
--- a/tests/ref/fate/rgb24-mkv
+++ b/tests/ref/fate/rgb24-mkv
@@ -1,4 +1,4 @@
-e85dcc076f7ebea460517c3255705fec *tests/data/fate/rgb24-mkv.matroska
+09a0bde5634a2a9d03bd640e13b96b7b *tests/data/fate/rgb24-mkv.matroska
 58213 tests/data/fate/rgb24-mkv.matroska
 #tb 0: 1/10
 #media_type 0: video
diff --git a/tests/ref/lavf/mka b/tests/ref/lavf/mka
index 6f9a56b..6fe5ca8 100644
--- a/tests/ref/lavf/mka
+++ b/tests/ref/lavf/mka
@@ -1,3 +1,3 @@
-b9c2d23dffba449cf1111f062b93c93f *tests/data/lavf/lavf.mka
-43588 tests/data/lavf/lavf.mka
+00381b9855734adb05bc96cfd2f3b6b2 *tests/data/lavf/lavf.mka
+43584 tests/data/lavf/lavf.mka
 tests/data/lavf/lavf.mka CRC=0x3a1da17e
diff --git a/tests/ref/lavf/mkv b/tests/ref/lavf/mkv
index 4626e7a..3212ab8 100644
--- a/tests/ref/lavf/mkv
+++ b/tests/ref/lavf/mkv
@@ -1,3 +1,3 @@
-1e6bddfa6a19e4c1a42401b5b0d556f0 *tests/data/lavf/lavf.mkv
-320461 tests/data/lavf/lavf.mkv
+68d084711bed456bc7ccda14c26c9246 *tests/data/lavf/lavf.mkv
+320457 tests/data/lavf/lavf.mkv
 tests/data/lavf/lavf.mkv CRC=0x5df3232f
diff --git a/tests/ref/lavf/mkv_attachment b/tests/ref/lavf/mkv_attachment
index 9af9e61..25d6140 100644
--- a/tests/ref/lavf/mkv_attachment
+++ b/tests/ref/lavf/mkv_attachment
@@ -1,3 +1,3 @@
-74beaeb8a0ddb24490b8c8de9ee8c54f *tests/data/lavf/lavf.mkv_attachment
-472611 tests/data/lavf/lavf.mkv_attachment
+7041f1831cc8c36dd8d2b63a7830bc08 *tests/data/lavf/lavf.mkv_attachment
+472607 tests/data/lavf/lavf.mkv_attachment
 tests/data/lavf/lavf.mkv_attachment CRC=0x5df3232f
diff --git a/tests/ref/seek/lavf-mkv b/tests/ref/seek/lavf-mkv
index ded131c..165c127 100644
--- a/tests/ref/seek/lavf-mkv
+++ b/tests/ref/seek/lavf-mkv
@@ -1,48 +1,48 @@
-ret: 0         st: 1 flags:1 dts:-0.011000 pts:-0.011000 pos:    697 size:   208 skip:481 discard:0
+ret: 0         st: 1 flags:1 dts:-0.011000 pts:-0.011000 pos:    693 size:   208 skip:481 discard:0
 ret: 0         st:-1 flags:0  ts:-1.000000
-ret: 0         st: 0 flags:1 dts: 0.000000 pts: 0.000000 pos:    913 size: 27837
+ret: 0         st: 0 flags:1 dts: 0.000000 pts: 0.000000 pos:    909 size: 27837
 ret: 0         st:-1 flags:1  ts: 1.894167
-ret: 0         st: 0 flags:1 dts: 0.960000 pts: 0.960000 pos: 292329 size: 27834
+ret: 0         st: 0 flags:1 dts: 0.960000 pts: 0.960000 pos: 292325 size: 27834
 ret: 0         st: 0 flags:0  ts: 0.788000
-ret: 0         st: 0 flags:1 dts: 0.960000 pts: 0.960000 pos: 292329 size: 27834
+ret: 0         st: 0 flags:1 dts: 0.960000 pts: 0.960000 pos: 292325 size: 27834
 ret: 0         st: 0 flags:1  ts:-0.317000
-ret: 0         st: 0 flags:1 dts: 0.000000 pts: 0.000000 pos:    913 size: 27837
+ret: 0         st: 0 flags:1 dts: 0.000000 pts: 0.000000 pos:    909 size: 27837
 ret:-1         st: 1 flags:0  ts: 2.577000
 ret: 0         st: 1 flags:1  ts: 1.471000
-ret: 0         st: 1 flags:1 dts: 0.982000 pts: 0.982000 pos: 320173 size:   209 skip:0 discard:347
+ret: 0         st: 1 flags:1 dts: 0.982000 pts: 0.982000 pos: 320169 size:   209 skip:0 discard:347
 ret: 0         st:-1 flags:0  ts: 0.365002
-ret: 0         st: 0 flags:1 dts: 0.480000 pts: 0.480000 pos: 146881 size: 27925
+ret: 0         st: 0 flags:1 dts: 0.480000 pts: 0.480000 pos: 146877 size: 27925
 ret: 0         st:-1 flags:1  ts:-0.740831
-ret: 0         st: 0 flags:1 dts: 0.000000 pts: 0.000000 pos:    913 size: 27837
+ret: 0         st: 0 flags:1 dts: 0.000000 pts: 0.000000 pos:    909 size: 27837
 ret:-1         st: 0 flags:0  ts: 2.153000
 ret: 0         st: 0 flags:1  ts: 1.048000
-ret: 0         st: 0 flags:1 dts: 0.960000 pts: 0.960000 pos: 292329 size: 27834
+ret: 0         st: 0 flags:1 dts: 0.960000 pts: 0.960000 pos: 292325 size: 27834
 ret: 0         st: 1 flags:0  ts:-0.058000
-ret: 0         st: 1 flags:1 dts:-0.011000 pts:-0.011000 pos:    697 size:   208 skip:481 discard:0
+ret: 0         st: 1 flags:1 dts:-0.011000 pts:-0.011000 pos:    693 size:   208 skip:481 discard:0
 ret: 0         st: 1 flags:1  ts: 2.836000
-ret: 0         st: 1 flags:1 dts: 0.982000 pts: 0.982000 pos: 320173 size:   209 skip:0 discard:347
+ret: 0         st: 1 flags:1 dts: 0.982000 pts: 0.982000 pos: 320169 size:   209 skip:0 discard:347
 ret:-1         st:-1 flags:0  ts: 1.730004
 ret: 0         st:-1 flags:1  ts: 0.624171
-ret: 0         st: 0 flags:1 dts: 0.480000 pts: 0.480000 pos: 146881 size: 27925
+ret: 0         st: 0 flags:1 dts: 0.480000 pts: 0.480000 pos: 146877 size: 27925
 ret: 0         st: 0 flags:0  ts:-0.482000
-ret: 0         st: 0 flags:1 dts: 0.000000 pts: 0.000000 pos:    913 size: 27837
+ret: 0         st: 0 flags:1 dts: 0.000000 pts: 0.000000 pos:    909 size: 27837
 ret: 0         st: 0 flags:1  ts: 2.413000
-ret: 0         st: 0 flags:1 dts: 0.960000 pts: 0.960000 pos: 292329 size: 27834
+ret: 0         st: 0 flags:1 dts: 0.960000 pts: 0.960000 pos: 292325 size: 27834
 ret:-1         st: 1 flags:0  ts: 1.307000
 ret: 0         st: 1 flags:1  ts: 0.201000
-ret: 0         st: 1 flags:1 dts:-0.011000 pts:-0.011000 pos:    697 size:   208 skip:481 discard:0
+ret: 0         st: 1 flags:1 dts:-0.011000 pts:-0.011000 pos:    693 size:   208 skip:481 discard:0
 ret: 0         st:-1 flags:0  ts:-0.904994
-ret: 0         st: 0 flags:1 dts: 0.000000 pts: 0.000000 pos:    913 size: 27837
+ret: 0         st: 0 flags:1 dts: 0.000000 pts: 0.000000 pos:    909 size: 27837
 ret: 0         st:-1 flags:1  ts: 1.989173
-ret: 0         st: 0 flags:1 dts: 0.960000 pts: 0.960000 pos: 292329 size: 27834
+ret: 0         st: 0 flags:1 dts: 0.960000 pts: 0.960000 pos: 292325 size: 27834
 ret: 0         st: 0 flags:0  ts: 0.883000
-ret: 0         st: 0 flags:1 dts: 0.960000 pts: 0.960000 pos: 292329 size: 27834
+ret: 0         st: 0 flags:1 dts: 0.960000 pts: 0.960000 pos: 292325 size: 27834
 ret: 0         st: 0 flags:1  ts:-0.222000
-ret: 0         st: 0 flags:1 dts: 0.000000 pts: 0.000000 pos:    913 size: 27837
+ret: 0         st: 0 flags:1 dts: 0.000000 pts: 0.000000 pos:    909 size: 27837
 ret:-1         st: 1 flags:0  ts: 2.672000
 ret: 0         st: 1 flags:1  ts: 1.566000
-ret: 0         st: 1 flags:1 dts: 0.982000 pts: 0.982000 pos: 320173 size:   209 skip:0 discard:347
+ret: 0         st: 1 flags:1 dts: 0.982000 pts: 0.982000 pos: 320169 size:   209 skip:0 discard:347
 ret: 0         st:-1 flags:0  ts: 0.460008
-ret: 0         st: 0 flags:1 dts: 0.480000 pts: 0.480000 pos: 146881 size: 27925
+ret: 0         st: 0 flags:1 dts: 0.480000 pts: 0.480000 pos: 146877 size: 27925
 ret: 0         st:-1 flags:1  ts:-0.645825
-ret: 0         st: 0 flags:1 dts: 0.000000 pts: 0.000000 pos:    913 size: 27837
+ret: 0         st: 0 flags:1 dts: 0.000000 pts: 0.000000 pos:    909 size: 27837
-- 
2.52.0

